### DISCUSSION TITLE
v9.1

### DISCUSSION THREAD

**Discussion by Mizux** [google/or-tools#2800] [Announces]
# Platform changes
* Ubuntu 16.04 LTS is deprecated (https://github.com/actions/virtual-environments/issues/3287).
* Add support for Debian 11 (Bullseye).
* Add support for Fedora 34.
* python: use `manylinux2014` image ([PEP 599](https://www.python.org/dev/peps/pep-0599/)).
* python: add support for aarch64 linux using `manylinux2014_aarch64` image.
* .Net: add .Net5 support.

# Dependencies Update
* abseil-cpp `20210324.1` -> `20210324.2`.
* Protobuf `3.15.8` -> `3.18.0`.
* SCIP `7.0.1` -> `master`.
* Googletest `1.8.0` -> `1.10.0`.
* python: use of `warning` in `cp_model.py` (#2530).
* python: absl-py `0.11` -> `0.13`.

# CMake
* Bump minimum version required 3.14 -> 3.15 (#2528).
* python: bump minimum required version 3.14 -> 3.18 (#2774).

# Make
* Make based build is deprecated please migrate to Bazel or CMake

# Java
* Improve robustness of the native library loader (#2742).
* Fix jvm GC crash when routing model or the constraint solver were disposed (#2091, #2466).
* Fix CP-SAT logging callback crash when using multiple workers (#2775).

# CP-SAT
* Improve robustness of the LNS code  (see #2525).
* Improve scheduling code: new factory methods to create fixed size intervals,
      new search heuristics, improved presolve and new linear cuts.
* Improve routing code: new dedicated LNS.
* Improve model checker. It is now more pedantic, especially w.r.t. potential overflows.
* Improve MIP code: better presolve and multiple improvements to the linear relaxation of MIP
      and CP models.
* Improve search diversity. When using more than 12 workers, add workers dedicated to
      improving the lower bound of the objective.
* Change to the parallelism code: by default, the solver will now use all available cores.
      Use the `num_search_parameters` to specify the level of parallelism.
* Deprecate `SearchAllSolutions` and `SolveWithSolutionCallback`.
* Python API: more pedantic checks when using `var == ...` or `var != ...` outside a `model.Add()` call.

<hr /><em>This discussion was created from the release <a href='https://github.com/google/or-tools/releases/tag/'>v9.1</a>.</em>

**Comment by CADBIMDeveloper** [comment]
Great news! CP-SAT solver becomes MUCH faster, thank you!

Only one question related to nuget package Google.OrTools.runtime.win-x64.9.1.9490. Could you check why it became so big now? ortools.lib is more than 380 Mb now. Pretty sure something went wrong...

**Reply by Mizux** [reply]
i'll take a look..., please notice now we target netcoreapp3.1 and net5.0
According to release page: https://github.com/google/or-tools/releases/tag/v9.1
Google.OrTools.runtime.win-x64.9.1.9490.nupkg 93.6 MB
note: linux and macOS are ~10MB
And: 
Google.OrTools.runtime.win-x64.9.0.9048.nupkg 10.9 MB

note2: For python wheel windows package are ~50MB in v9.0 AND v9.1

**Reply by btzdnl** [reply]
The `.nupkg` is “only” ~94 MiB, that’s correct. However, the unpacked files are an incredible ~523 MiB. Surely that can’t be right? That’s almost 20 times the previous size. It quickly fills up disk space on CI workers.

There’s a ton of `.lib` files there. Aren’t those static libraries? I don’t see how .NET could use any of them.

**Reply by Mizux** [reply]
```sh
[^u^]─mizux@nuc10i7 %for i in *.nupkg; do unzip -l $i; done
Archive:  google.ortools.runtime.win-x64.9.0.9048.nupkg
  Length      Date    Time    Name
---------  ---------- -----   ----
      524  2021-04-29 20:53   _rels/.rels
     1166  2021-04-29 20:53   Google.OrTools.runtime.win-x64.nuspec
 28535296  2021-04-29 20:53   runtimes/win-x64/native/google-ortools-native.dll
      910  2021-02-25 10:02   build/net45/Google.OrTools.runtime.win-x64.targets
     6141  2021-02-25 10:02   content/orLogo.png
      595  2021-04-29 20:53   [Content_Types].xml
      768  2021-04-29 20:53   package/services/metadata/core-properties/1e19ac8af6074a7797aafa8fc3eff114.psmdcp
     9476  2021-04-30 05:40   .signature.p7s
---------                     -------
 28554876                     8 files
Archive:  google.ortools.runtime.win-x64.9.1.9490.nupkg
  Length      Date    Time    Name
---------  ---------- -----   ----
      524  2021-10-01 03:53   _rels/.rels
     1043  2021-10-01 03:53   Google.OrTools.runtime.win-x64.nuspec
 28944384  2021-10-01 03:53   runtimes/win-x64/native/google-ortools-native.dll
401698176  2021-10-01 03:49   runtimes/win-x64/native/ortools.lib
    15610  2021-10-01 03:18   runtimes/win-x64/native/absl_bad_any_cast_impl.lib
     1204  2021-10-01 03:17   runtimes/win-x64/native/absl_bad_optional_access.lib
     1202  2021-10-01 03:17   runtimes/win-x64/native/absl_bad_variant_access.lib
    70040  2021-10-01 03:16   runtimes/win-x64/native/absl_base.lib
    17648  2021-10-01 03:17   runtimes/win-x64/native/absl_city.lib
   179804  2021-10-01 03:17   runtimes/win-x64/native/absl_civil_time.lib
  1559544  2021-10-01 03:17   runtimes/win-x64/native/absl_cord.lib
     3418  2021-10-01 03:16   runtimes/win-x64/native/absl_debugging_internal.lib
   134074  2021-10-01 03:16   runtimes/win-x64/native/absl_demangle_internal.lib
    10394  2021-10-01 03:17   runtimes/win-x64/native/absl_examine_stack.lib
     7016  2021-10-01 03:16   runtimes/win-x64/native/absl_exponential_biased.lib
    22432  2021-10-01 03:17   runtimes/win-x64/native/absl_failure_signal_handler.lib
     2382  2021-10-01 03:17   runtimes/win-x64/native/absl_flags.lib
     2776  2021-10-01 03:17   runtimes/win-x64/native/absl_flags_commandlineflag.lib
     4674  2021-10-01 03:17   runtimes/win-x64/native/absl_flags_commandlineflag_internal.lib
   270404  2021-10-01 03:17   runtimes/win-x64/native/absl_flags_config.lib
   258776  2021-10-01 03:17   runtimes/win-x64/native/absl_flags_internal.lib
   525112  2021-10-01 03:17   runtimes/win-x64/native/absl_flags_marshalling.lib
   948170  2021-10-01 03:18   runtimes/win-x64/native/absl_flags_parse.lib
     9308  2021-10-01 03:17   runtimes/win-x64/native/absl_flags_private_handle_accessor.lib
    82524  2021-10-01 03:17   runtimes/win-x64/native/absl_flags_program_name.lib
   916692  2021-10-01 03:17   runtimes/win-x64/native/absl_flags_reflection.lib
    68586  2021-10-01 03:17   runtimes/win-x64/native/absl_flags_usage.lib
  1144932  2021-10-01 03:17   runtimes/win-x64/native/absl_flags_usage_internal.lib
   131026  2021-10-01 03:17   runtimes/win-x64/native/absl_graphcycles_internal.lib
    15524  2021-10-01 03:17   runtimes/win-x64/native/absl_hash.lib
    79862  2021-10-01 03:17   runtimes/win-x64/native/absl_hashtablez_sampler.lib
   185352  2021-10-01 03:16   runtimes/win-x64/native/absl_int128.lib
     2962  2021-10-01 03:17   runtimes/win-x64/native/absl_leak_check.lib
     1192  2021-10-01 03:17   runtimes/win-x64/native/absl_leak_check_disable.lib
    15838  2021-10-01 03:16   runtimes/win-x64/native/absl_log_severity.lib
    59194  2021-10-01 03:16   runtimes/win-x64/native/absl_malloc_internal.lib
     3582  2021-10-01 03:16   runtimes/win-x64/native/absl_periodic_sampler.lib
   102226  2021-10-01 03:18   runtimes/win-x64/native/absl_random_distributions.lib
   111162  2021-10-01 03:18   runtimes/win-x64/native/absl_random_internal_distribution_test_util.lib
     6032  2021-10-01 03:18   runtimes/win-x64/native/absl_random_internal_platform.lib
    73018  2021-10-01 03:18   runtimes/win-x64/native/absl_random_internal_pool_urbg.lib
     6618  2021-10-01 03:18   runtimes/win-x64/native/absl_random_internal_randen.lib
     1956  2021-10-01 03:18   runtimes/win-x64/native/absl_random_internal_randen_hwaes.lib
     7712  2021-10-01 03:18   runtimes/win-x64/native/absl_random_internal_randen_hwaes_impl.lib
    20480  2021-10-01 03:18   runtimes/win-x64/native/absl_random_internal_randen_slow.lib
    16304  2021-10-01 03:18   runtimes/win-x64/native/absl_random_internal_seed_material.lib
    11946  2021-10-01 03:18   runtimes/win-x64/native/absl_random_seed_gen_exception.lib
    60882  2021-10-01 03:18   runtimes/win-x64/native/absl_random_seed_sequences.lib
     5102  2021-10-01 03:17   runtimes/win-x64/native/absl_raw_hash_set.lib
    30754  2021-10-01 03:16   runtimes/win-x64/native/absl_raw_logging_internal.lib
    67452  2021-10-01 03:16   runtimes/win-x64/native/absl_scoped_set_env.lib
    12078  2021-10-01 03:16   runtimes/win-x64/native/absl_spinlock_wait.lib
    15446  2021-10-01 03:16   runtimes/win-x64/native/absl_stacktrace.lib
   528138  2021-10-01 03:18   runtimes/win-x64/native/absl_status.lib
    92264  2021-10-01 03:18   runtimes/win-x64/native/absl_statusor.lib
    72438  2021-10-01 03:16   runtimes/win-x64/native/absl_strerror.lib
   895758  2021-10-01 03:16   runtimes/win-x64/native/absl_strings.lib
    52310  2021-10-01 03:16   runtimes/win-x64/native/absl_strings_internal.lib
  1017206  2021-10-01 03:17   runtimes/win-x64/native/absl_str_format_internal.lib
     5576  2021-10-01 03:17   runtimes/win-x64/native/absl_symbolize.lib
   297256  2021-10-01 03:17   runtimes/win-x64/native/absl_synchronization.lib
   115260  2021-10-01 03:16   runtimes/win-x64/native/absl_throw_delegate.lib
   702964  2021-10-01 03:17   runtimes/win-x64/native/absl_time.lib
  1659278  2021-10-01 03:17   runtimes/win-x64/native/absl_time_zone.lib
     5962  2021-10-01 03:17   runtimes/win-x64/native/absl_wyhash.lib
  5334372  2021-10-01 03:22   runtimes/win-x64/native/libprotobuf-lite.lib
 37195774  2021-10-01 03:25   runtimes/win-x64/native/libprotobuf.lib
 41453408  2021-10-01 03:27   runtimes/win-x64/native/libprotoc.lib
 21514082  2021-10-01 03:22   runtimes/win-x64/native/libscip.lib
   353212  2021-10-01 03:16   runtimes/win-x64/native/zlib.lib
     6141  2021-02-25 10:02   content/orLogo.png
      591  2021-10-01 03:54   [Content_Types].xml
      769  2021-10-01 03:54   package/services/metadata/core-properties/64e4a62b2f3a4dd9920658a263e0fbe6.psmdcp
     9475  2021-10-01 09:05   .signature.p7s
---------                     -------
549256783                     74 files
```
So in v9.1 we captured all .lib files which should have been uneeded

**Reply by CADBIMDeveloper** [reply]
@Mizux , could you also get back a project target file in v.9.2 for .Net 4.*, Something like it was in 9.0, please:
`  <Target Name="EnsureNuGetPackageBuildImports" BeforeTargets="PrepareForBuild">
    <PropertyGroup>
      <ErrorText>This project references NuGet package(s) that are missing on this computer. Use NuGet Package Restore to download them.  For more information, see http://go.microsoft.com/fwlink/?LinkID=322105. The missing file is {0}.</ErrorText>
    </PropertyGroup>
    <Error Condition="!Exists('..\..\packages\Google.OrTools.runtime.win-x64.9.0.9048\build\net45\Google.OrTools.runtime.win-x64.targets')" Text="$([System.String]::Format('$(ErrorText)', '..\..\packages\Google.OrTools.runtime.win-x64.9.0.9048\build\net45\Google.OrTools.runtime.win-x64.targets'))" />
  </Target>`

**Comment by mslazynski** [comment]
Thanks for the great work!

Could you maybe elaborate a bit more on this change:  "Improve routing code: new dedicated LNS."?
Does it mean that CP-SAT could now effectively replace an old CP in the routing solver? 
Or do I misread the thing completely?

**Reply by Mizux** [reply]
I mean: `git diff v9.0...v9.1 -- ortools/sat/cp_model_lns.cc ortools/sat/cp_model_lns.h`
Otherwise CP-SAT won't replace the actual CP based routing solver (IIRC CP still outperform for search space exploration while CP-SAT is suitable for scheduling breaks) it seems there are complementary than a pure replacement.

There is still some experimental test of using CP-SAT as a routing replacement for simple example here:
For a pure SAT routing you can take a look at:
* https://github.com/google/or-tools/blob/stable/examples/python/tsp_sat.py
* https://github.com/google/or-tools/blob/stable/examples/python/prize_collecting_tsp_sat.py
